misc/bstr.c

```diff
@@ -454,3 +454,27 @@ struct bstr bstr_get_ext(struct bstr s)
         return (struct bstr){NULL, 0};
     return bstr_splice(s, dotpos + 1, s.len);
 }
+
+struct bstr bstr_decode_hex(void *talloc_ctx, struct bstr hex)
+{
+    char *out = talloc_array(talloc_ctx, char, hex.len);
+    int pos = 0;
+
+    // Read out all hex chars as individual nibbles
+    for (int i = 0; i < hex.len; i++) {
+        unsigned char in = hex.start[i];
+        if (in >= '0' && in <= '9')
+            out[pos++] = in - '0';
+        if (in >= 'a' && in <= 'f')
+            out[pos++] = in - 'a' + 10;
+        if (in >= 'A' && in <= 'F')
+            out[pos++] = in - 'A' + 10;
+    }
+
+    // Go through the output and recombine all of the nibbles in-place
+    int len = pos / 2;
+    for (int i = 0; i < len; i++)
+        out[i] = (out[2*i+0] << 4) | out[2*i+1];
```

Comment:
> I'd probably treat it as invalid.

So you want the validity of user shaders to (arbitrarily) depend on what width you decide to wrap your hex representation to? I.e. wrapping to 78 or 80 columns would work, but wrapping to 79 would be invalid?